### DISCUSSION TITLE
What is the best way to check if element is present for controlling flow of automation script?

### DISCUSSION THREAD

**Discussion by edwin-hendriks** [selenide/selenide#1893] [Q&A]
Hi dear Selenide team,

Thanks for Selenide. Great tool and may yet become my goto test automation framework. 

Perhaps you can help me with the following. I am testing a Vaadin based application. In it I need to be able to check for an element to be present/visible. If so do some extra clicks. If not I can skip that or do other clicks. So no exception should occur. I tried several options. Non really worked so far the most logical option for me was to create a (Vaadin)IsPresent method that would try shortly to check if an element is visible. If not it will cause an Exception which is caught by the catch part. 

```java
	protected boolean vIsPresent(String xpath) {
		try {
			$x(xpath).shouldBe(visible, Duration.ofMillis(500));
                        return true;
		} catch (Exception e) {
			return false;
		}
	}
```
This is not working because this piece of code still causes an exception. So my question to you would be, what is the best approach to resolve this.

**Comment by asolntsev** [comment]
The correct answer is: a good test should not use IFs. 
Instead, you should have 2 tests: one where the element is present, and another where it's not. 
Otherwise one of these cases might not be covered. 

More details here: https://selenide.org/2019/12/02/advent-calendar-how-to-abuse-selenide/

**Reply by edwin-hendriks** [reply]
Thanks for the answer Andrei. I read the details and I agree. However, one big problem: I am testing/automating a standard out of the box app from an external vendor over which I have no control whatsoever. I really need a kind-a-logic in which I need to test if something is there, and if so do X and if not do Y. 

Could you advise the best approach?

**Reply by asolntsev** [reply]
There was an answer in the article mentioned  before.

**Reply by edwin-hendriks** [reply]
I have read the article and I could not find a solution I can implement or achieve. Are you referring to the solutions under **What can you do?**? If so please let me know.

Sadly I am not able to choose any of these options and catching the `AssertionError` seems the best of the only option that I have available. Don't blame me, but blame the vendor of the tool I need to test/automate.

**Reply by asolntsev** [reply]
Yes, I meant catching the `AssertionError`. 

P.S. I don't blame anybody. You asked me to advise the best approach, so I am trying to advice the best approach.

**Comment by edwin-hendriks** [comment]
Thanks. We are currently going for that solution.

Edwin Hendriks
+31625014678

Op vr 15 jul. 2022 15:15 schreef Andrei Solntsev ***@***.***>:
